### ISSUE TITLE
Twig text templates get HTML to text conversion

### ISSUE THREAD

**Issue by cruftex** [PrestaShop/PrestaShop#32971]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When creating templates for text only mails to HTML to text transformation is done when generating templates.

### Expected behavior

Keep text files as is, since input and output is text.

### Steps to reproduce

1. Create a layout in `mails/themes/test/core/hello.txt.twig`, with

````
Expect this not to be removed:
<br>
And also result contains new lines
````

2. Generate Emails with email theme `test` and overwrite core templates
3. Inspect result at `mails/en/hello.txt`

### PrestaShop version(s) where the bug happened

8.0.4

### PHP version(s) where the bug happened

8.1

### If your bug is related to a module, specify its name and its version

n/a

### Your company or customer's name goes here (if applicable).

headissue GmbH

**Comment by AureRita** [comment]
Hi @cruftex 

Thank you for your report, I tried to reproduce your issue as you can see : 

[Untitled_ Jun 22 2023 11_42 AM.webm](https://github.com/PrestaShop/PrestaShop/assets/100699445/6a3943a9-f8d7-4642-8dee-627c701cd6b6)

If I understand well, in this case, I should have "< br >" showed ? 
or should I see another issue ? 

Waiting for your feedback

**Comment by cruftex** [comment]
@AureRita:
Thanks for reproducing!

The "\<br\>" needs to be in the expected output. 

If the "\<br\>" is missing that is the indication that the HTML to text transformation did run.

There also need to be at least two newlines. The HTML to text transformation is converting new lines to space and converting the "\<br\>" to newline.

**Comment by AureRita** [comment]
Thank you for the details, so I reproduce the issue with PrestaShop version1.7.8.9, I'll add this to the backlog so it can be fixed.

Thank you for your PR :rocket: 

Thanks

**Comment by jolelievre** [comment]
Hi @cruftex 

I don't understand your request, the text template is cleaned on purpose that's part of the feature. The text part of the mail is supposed to be pure text and not contain some HTML. If you want to use HTML tags like <br> then it should be used inside the HTML template, not the text one.

The text part is used by mail readers that can't interpret HTML, so it won't recognize <br> as an end of the line, simply as some text `<br>`. As for the conversions you mentioned it's also on purpose, when no text template is available the text part is automatically generated based on the content of the HTML content, we can't keep all the EOL as is from HTML input since they often contain many once tags are stripped, so just like in a regular HTML display EOL are interpreted as regular spaces and only <br> which are explicit HTML EOL are replaced as such.

One thing that could maybe be improved though from your example is to detect when there is an explicit text template, in this case the clean/strip tags should still be applied no matter what. However we could maybe keep the EOL as is, but only from explicit txt twig templates. The your text templates would look like this:

```txt
Expect this not to be removed:

And also result contains new lines
```

**Comment by cruftex** [comment]
@jolelievre
Why do you want to interpret it as HTML and convert it to text if it is text already?

I managed well by adding "&lt;br>" everywhere, however then I failed to make a proper email footer. An RFC compliant signature, looks like this:

````
-- 
Signature
````

There needs to be a trailing space after "--". Unfortunately, the cleanup strips trailing spaces. I know, that is pretty old school, but all mail agents support this properly.

Probably leading spaces are stripped also, so you cannot indent lines.

Why am I forced to write HTML entities for the characters &, < and > although it is a text mail?

To make proper text mails, I need it to preserve the characters as I put them in. I thought that is why there is the separate `.txt.twig` layout file.

I am very thankful the feature is there.

I doubt that anybody else is actually using it, since there is no documentation and example yet. But, when I Google for it, I find this issue now :)

**Comment by cruftex** [comment]
yes

**Comment by boo-code** [comment]
@jolelievre, I think the disagreement here is narrower than it looks, and `HTMLToTextTransformation`'s own docblock resolves it:

> It is **especially useful when no txt layout is defined** and the renderer uses the html layout as a base. This transformation then removes any html tags but keep the raw information.

You are right that the text part must be pure text, and right that when the text part is derived from the HTML template it has to be converted. That is the case the class describes. The other case is a theme that ships a `.txt.twig`: there the author has already written text, and the conversion has nothing to remove and only damage to do. `MailTemplateTwigRenderer::render()` distinguishes the two itself — `getTxtPath()` when it is set, `getHtmlPath()` when it is not — so the two paths are already separate everywhere except in which transformations get applied.

Measured on 9.2, running the transformation over @cruftex's sample:

```
INPUT   "Expect this not to be removed:\n<br>\nAnd also result contains new lines\n\n-- \nSignature\n"
OUTPUT  "Expect this not to be removed:\nAnd also result contains new lines -- Signature\n"

br tag kept:           NO
blank line kept:       NO
signature '-- ' kept:  NO
```

The last line is the one I would weigh most. An RFC 3676 signature delimiter is `-- ` with a trailing space, and mail clients stop treating what follows as a signature once it is stripped — so a theme cannot produce a correct plain-text signature at all today.

**On the scope of changing this:** `find mails/themes -name '*.txt.twig'` returns **0**, against **94** `*.html.twig`. No core theme ships a text layout, so every core mail keeps taking the fallback path and its text part is generated exactly as it is now. The only thing that changes is a theme that deliberately authors a `.txt.twig`.

**On #32973.** @cruftex, the reason I would not merge it as it stands is the detection:

```php
if (strlen($templateContent) > 0 && substr($templateContent, 0, 1) != '<') {
    return $templateContent;
}
```

That reads the first character of the rendered output. An HTML layout whose output starts with a comment, a newline or a Twig-emitted space is taken for text, and a text layout that legitimately starts with `<` is still converted. It also removes the `LoaderError` to `FileNotFoundException` handling and 39 lines of that class's tests, which are unrelated and would draw review onto the wrong thing.

The renderer already knows which layout it used, so the decision does not need guessing from content. In `getMailLayoutTransformations()`, which already filters transformations and already carries one special case:

```php
if ($transformation instanceof HTMLToTextTransformation && !empty($mailLayout->getTxtPath())) {
    continue;
}
```

I have that on a branch against `9.2.x` with two unit tests — one asserting an authored text layout comes through untouched, one asserting the HTML fallback still converts. With the guard reverted, exactly the first fails and the second, plus the seven existing tests in that class, still pass.

Not opening it, since #32973 is @cruftex's and this is a two-line variation on their idea. Happy to hand it over, or to push it if that is easier.

**Comment by cruftex** [comment]
@boo-code 
> That reads the first character of the rendered output. An HTML layout whose output starts with a comment, a newline or a Twig-emitted space is taken for text, and a text layout that legitimately starts with < is still converted. It also removes the LoaderError to FileNotFoundException handling and 39 lines of that class's tests, which are unrelated and would draw review onto the wrong thing.

Agreed. If you can solve it in a cleaner way, please go ahead and do a new PR.

My current PR is hacky, however, it works for us at the moment and maybe others as well that want to do text mails. So its good that its there, while maybe not being in a good shape to go into the main code.
